### DISCUSSION TITLE
Upgrade policy

### DISCUSSION THREAD

**Discussion by waiting-for-dev** [solidusio/solidus#4045] [New Features or Ideas]
We talked with @kennyadsl and @cpfergus1 about ideas to provide users with a clear path to upgrade major releases of Solidus. Here it's a wrap-up so that we can start a discussion with the broader community.

We all agreed that the _future compatible_ model we're already following is a good idea. That's, having the last minor version of a major branch deprecating through warnings all the code that won't work in the next major version anymore.

We also have changes in behavior that can support backward compatibility through a setting. At this point, what we're doing is adding a new setting with a default value that keeps the deprecated behavior. However, when a new Solidus app is created, an initializer is generated, setting the configuration to the new recommended value. This way, only fresh installations will use it by default. See https://github.com/solidusio/solidus/pull/3749 as an example.

This is similar to what Rails are doing. They also add a setting in which the default value keeps the deprecated behavior. For new applications, instead of setting the recommended value in an initializer, they call the [`load_defaults`](https://api.rubyonrails.org/classes/Rails/Application/Configuration.html#method-i-load_defaults) method with the latest Rails version. However, they also provide some guidance to update existing installations. The `rails app:update` task creates an initializer where all the new defaults are commented. Users can activate them one by one and, once there're no more items to enable, they can update the `load_defaults` call and remove the initializer altogether.

Both approaches have a very inconvenient drawback. They leave the core code with a deprecated default value. It makes the code reflect the history that leads to the current snapshot, making things more difficult to understand.

We discussed an alternative approach is to have an update task that creates a `solidus_{OLDER_VERSION}_defaults` initializer instead. In this way, we would also keep backward compatibility behavior while our core code would reflect the current state of recommendations. As in the Rails approach, users would still be able to update step-by-step. Eventually, they could also remove the initializer altogether.

This workflow also comes with its drawbacks. To begin with, we'd differing from what Rails is doing. More importantly, users who upgrade the application without running the update task would end up without the initializer and, therefore, breaking their application. On Rails, a blind update leaves you on the safe side. On top of that, `load_defaults` groups settings in chunks, which can be useful for debugging. However, reading the docs for a major upgrade is a very minimum requirement for a developer, and having less complexity on Solidus code is a great benefit.

On Rails, the `rails app:update` does more than that. The task tries to update what is needed on the code magically. [It does so](https://github.com/rails/rails/blob/a57c6eee6db189c9821744eff19639262d48c841/railties/lib/rails/app_updater.rb) in attempting to run `rails new` with some tweaks on top of the old application. We talked that it's probably a bad idea because it's brittle and encourages users to modify the code blindly. For these kinds of changes, a good upgrade guide is the only reliable option.

Ensuring that we provide an excellent upgrade guide is paramount to help users in the process. We considered some options:

- Enforcing contributors to modify the Changelog if needed. It was a policy we had in the past, but it discouraged some people or led to poor descriptions. However, it's easier to be consistent.
- Break the no-push-to-master policy to allow our own updates to the Changelog after we merge a PR. It lowers the barrier for contributors, but it's easier to be laxer in the implementation.
- Modify the template that is rendered during a PR submission and add a new section *Upgrading instructions*. Then use some automated tool to extract the content and generate a Changelog or Upgrade guide. As Solidus members can edit the description of a PR, it would allow us to back up users who don't write it themselves or improve it. We could investigate a workflow with Github actions to enforce consistency. If that's possible, we could block the merge until something is added to the section unless it's a keyword like "Nothing", or only block when a label like "breaking-change" is added to the issue.

Thoughts?

**Comment by aldesantis** [comment]
Thanks for the detailed write-up, @waiting-for-dev! It sounds like we're trying to tackle the problem both with code and with better changelog practices, which I think is the right way to go. I'll try to leave my thoughts on some of your ideas below.

> Both approaches have a very inconvenient drawback. They leave the core code with a deprecated default value. It makes the code reflect the history that leads to the current snapshot, making things more difficult to understand.

This doesn't seem so annoying to me. If we want to avoid any confusion stemming from the deprecated defaults, we could consider simply adding a comment explaining that the default value is actually deprecated and that it's going to replaced in the next major version (or that the configuration option will be removed altogether), so that devs inspecting the code have a clear idea of what's going on.

An alternative, more complicated (but cooler!) approach could be to do this in code, e.g.

```ruby
preference :address_requires_phone, :boolean, defaults: {
  '2.10' => true,
  '2.11' => false,
}
```

This would make it clear that the setting has changed over time, and it would also allow us to implement some automated tooling on top of it (e.g., to generate an initializer with the new configuration options since your previous Solidus version). However, I don't think the effort here justifies the advantages.

When it comes to configuration management, a `load_defaults` mechanism such as Rails', in combination with an update task that generates an initializer for the new options, seems the most sensible thing to me: it's extremely simple and yet very effective at making upgrades easier. I also like the idea of generating an initializer for the older version, although I think there's a lot of value in sticking to what Rails does: it will make the experience of upgrading Solidus much more consistent and nicer for all Rails developers.

As for the changelog, how about having a core team member always update the changelog in the original PR before it's merged? GitHub allows you to update someone else's PR if you enable the feature, and this seems like the perfect use case for doing it. I don't like the idea of pushing to master because it's potentially dangerous and makes reverts (and reconstructing the Git history) harder.

**Reply by waiting-for-dev** [reply]
> GitHub allows you to update someone else's PR if you enable the feature, and this seems like the perfect use case for doing it. I don't like the idea of pushing to master because it's potentially dangerous and makes reverts (and reconstructing the Git history) harder.

Just a clarification on this. It's the contributor that can enable/disable the allowing of upstream maintainers to make changes. However, it's on by default when creating a PR, so probably we would not find problems with it and, if needed, we can ask them to enable it once the PR has been created. I think it's a good idea, although it may be a bit of a burden to do it every time.

![screenshot-github com-2021 05 06-05_52_20](https://user-images.githubusercontent.com/52650/117239859-d8449000-ae2f-11eb-9ab0-ac97d209de96.png)

**Reply by waiting-for-dev** [reply]
> An alternative, more complicated (but cooler!) approach could be to do this in code, e.g.
> 
> ```ruby
> preference :address_requires_phone, :boolean, defaults: {
>   '2.10' => true,
>   '2.11' => false,
> }
> ```

We decided to do a spike about it. Please, take a look at https://github.com/solidusio/solidus/pull/4064

**Comment by jarednorman** [comment]
Had a little internal discussion around this with @adammathys, @forkata, @benjaminwil, and @MadelineCollier.

We're basically all on the same page: the `load_defaults` mechanism from Rails and the process of having a generator that creates an initializer with the new defaults commented out to ease upgrading is the best option. In general, following the same upgrade process as Rails should be more familiar to people. One less new thing to learn when digging into Solidus seems like a win.
